overlord/fdestate/handlers.go

```diff
@@ -19,47 +19,181 @@
 
 import (
-	"github.com/snapcore/snapd/overlord/state"
+	"fmt"
+
 	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/secboot"
 )
 
-func (m *FDEManager) doAddRecoveryKeys(t *state.Task, tomb *tomb.Tomb) error {
-	// TODO:FDEM: implement recovery key addition, this is currently only a
-	// mock task for testing.
-
-	// TODO:FDEM:
-	//   - this might be a re-run, make task idempotent to be reselient to
-	//     abrupt reboot/shutdown.
-	//   - important to detect absence of recovery key ID and do cleanup
-	//     of added key slots on re-run and returning an error.
-	//   - distinguish between errors (undo) and pure-reboots (re-run).
-	//   - conflict detection for key slot tasks is important because it
-	//     reduces the possible states we could end up in.
+var (
+	secbootAddContainerRecoveryKey = secboot.AddContainerRecoveryKey
+	secbootDeleteContainerKey      = secboot.DeleteContainerKey
+	secbootRenameContainerKey      = secboot.RenameContainerKey
+)
+
+func (m *FDEManager) doAddRecoveryKeys(t *state.Task, tomb *tomb.Tomb) (err error) {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	var keyslotRefs []KeyslotRef
+	if err := t.Get("keyslots", &keyslotRefs); err != nil {
+		return err
+	}
+
+	var recoveryKeyID string
+	if err := t.Get("recovery-key-id", &recoveryKeyID); err != nil {
+		return err
+	}
+
+	// XXX: unlock state and let conflict detection handle the rest?
+
+	containers, err := m.GetEncryptedContainers()
+	if err != nil {
+		return err
+	}
+	containerDevicePath := make(map[string]string, len(containers))
+	for _, container := range containers {
+		containerDevicePath[container.ContainerRole()] = container.DevPath()
+	}
+
+	// IMPORTANT: this clean up must be declared as early as possible
+	// to account for real errors and potential re-runs (that will fail
+	// anyway as soon as the recovery key ID is reused).
+	defer func() {
+		if err == nil {
+			return
+		}
+		for _, keyslotRef := range keyslotRefs {
+			devicePath := containerDevicePath[keyslotRef.ContainerRole]
+			if err := secbootDeleteContainerKey(devicePath, keyslotRef.Name); err != nil {
+				// best effort deletion, log errors only
+				logger.Noticef("failed to delete %s during clean up: %v", keyslotRef.String(), err)
+			}
+		}
+	}()
+
+	currentKeyslots, missing, err := m.GetKeyslots(keyslotRefs)
+	if err != nil {
+		return fmt.Errorf("failed to find key slots: %v", err)
+	}
+	if len(missing) == 0 {
+		// this could be re-run and all key slots were already added, do nothing
+		return nil
+	}
+	if len(currentKeyslots) != 0 {
+		return &keyslotsAlreadyExistsError{keyslots: currentKeyslots}
+	}
+
+	rkey, err := m.getRecoveryKey(recoveryKeyID)
+	if err != nil {
+		// most likely a re-run as keys expire after first use and
+		// clean up is needed.
+		return fmt.Errorf("failed to find recovery key with id %q: %v", recoveryKeyID, err)
+	}
+
+	for _, keyslotRef := range keyslotRefs {
```

Comment:
should this iterate over missing?

Reply:
yes, I switched to iterate over missing while addressing some of the review comments above. it should be updated now.